community/index/src/main/java/org/neo4j/index/internal/gbptree/CrashGenCleaner.java

```diff
@@ -0,0 +1,245 @@
+        this.stableGeneration = stableGeneration;
+        this.unstableGeneration = unstableGeneration;
+        this.monitor = monitor;
+        this.maxKeyCount = treeNode.internalMaxKeyCount();
+    }
+
+    // === Methods about the execution and threading ===
+
+    public void clean() throws IOException
+    {
+        assert unstableGeneration > stableGeneration;
+        assert unstableGeneration - stableGeneration > 1;
+
+        long startTime = currentTimeMillis();
+        int threads = Runtime.getRuntime().availableProcessors();
+        ExecutorService executor = Executors.newFixedThreadPool( threads );
+        AtomicLong nextId = new AtomicLong( lowTreeNodeId);
+        AtomicReference<Throwable> error = new AtomicReference<>();
+        AtomicInteger cleanedPointers = new AtomicInteger();
+        for ( int i = 0; i < threads; i++ )
+        {
+            executor.submit( cleaner( nextId, error, cleanedPointers ) );
+        }
+        executor.shutdown();
+
+        try
+        {
+            long lastProgression = nextId.get();
+            // Have max no-progress-timeout quite high to be able to cope with huge
+            // I/O congestion spikes w/o failing in vain.
+            while ( !executor.awaitTermination( 30, SECONDS ) )
+            {
+                if ( lastProgression == nextId.get() )
+                {
+                    // No progression at all, abort?
+                    error.compareAndSet( null, new IOException( "No progress, so forcing abort" ) );
+                }
+                lastProgression = nextId.get();
+            }
+        }
+        catch ( InterruptedException e )
+        {
+            Thread.currentThread().interrupt();
+        }
+
+        Throwable finalError = error.get();
+        if ( finalError != null )
+        {
+            throw launderedException( IOException.class, finalError );
+        }
+
+        long endTime = currentTimeMillis();
+        monitor.recoveryCompleted( highTreeNodeId - lowTreeNodeId, cleanedPointers.get(), endTime - startTime );
+    }
+
+    private Runnable cleaner( AtomicLong nextId, AtomicReference<Throwable> error, AtomicInteger cleanedPointers )
+    {
+        return () ->
+        {
+            try ( PageCursor cursor = pagedFile.io( 0, PagedFile.PF_SHARED_READ_LOCK );
+                    PageCursor writeCursor = pagedFile.io( 0, PagedFile.PF_SHARED_WRITE_LOCK ) )
+            {
+                while ( nextId.get() < highTreeNodeId )
+                {
+                    // Do a batch of max 1000
+                    int batchCount = 0;
+                    for ( ; batchCount < 1_000; batchCount++ )
+                    {
+                        long id = nextId.getAndIncrement();
+                        if ( id >= highTreeNodeId )
+                        {
+                            break;
+                        }
+
+                        // TODO: we should try to figure out if the disk backing this paged file
```

Comment:
i don't think this comment is useful so far as a comment since we will never gonna come to this place again and introduce check for SSD or anything else like that, if we want to introduce something like that we should have a card instead